filters/shedder/admission_test.go

```diff
@@ -0,0 +1,199 @@
+		pExpectedAdmissionShedding: 0.1,
+	}, {
+		msg:                        "small error rate and bigger than threshold will block some traffic",
+		mode:                       "active",
+		d:                          10 * time.Millisecond,
+		windowsize:                 5,
+		minRequests:                10,
+		successThreshold:           0.9,
+		maxrejectprobability:       0.95,
+		aggression:                 1.0,
+		N:                          10000,
+		pBackendErr:                0.2,
+		pExpectedAdmissionShedding: 0.1,
+	}, {
+		msg:                        "medium error rate and bigger than threshold will block some traffic",
+		mode:                       "active",
+		d:                          10 * time.Millisecond,
+		windowsize:                 5,
+		minRequests:                10,
+		successThreshold:           0.9,
+		maxrejectprobability:       0.95,
+		aggression:                 1.0,
+		N:                          10000,
+		pBackendErr:                0.5,
+		pExpectedAdmissionShedding: 0.615,
+	}, {
+		msg:                        "large error rate and bigger than threshold will block some traffic",
+		mode:                       "active",
+		d:                          10 * time.Millisecond,
+		windowsize:                 5,
+		minRequests:                10,
+		successThreshold:           0.9,
+		maxrejectprobability:       0.95,
+		aggression:                 1.0,
+		N:                          10000,
+		pBackendErr:                0.8,
+		pExpectedAdmissionShedding: 0.91,
+	}} {
+		t.Run(ti.msg, func(t *testing.T) {
+			backend := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				p := rand.Float64()
+				if p < ti.pBackendErr {
+					w.WriteHeader(http.StatusInternalServerError)
+				} else {
+					w.WriteHeader(http.StatusOK)
+				}
+			}))
+
+			spec := NewAdmissionControl(Options{})
+			args := make([]interface{}, 0, 6)
+			args = append(args, "testmetric", ti.mode, ti.d.String(), ti.windowsize, ti.minRequests, ti.successThreshold, ti.maxrejectprobability, ti.aggression)
+			_, err := spec.CreateFilter(args)
+			if err != nil {
+				t.Logf("args: %+v", args...)
+				t.Fatalf("error creating filter: %v", err)
+				return
+			}
+
+			fr := make(filters.Registry)
+			fr.Register(spec)
+			r := &eskip.Route{Filters: []*eskip.Filter{{Name: spec.Name(), Args: args}}, Backend: backend.URL}
+
+			proxy := proxytest.New(fr, r)
+			reqURL, err := url.Parse(proxy.URL)
+			if err != nil {
+				t.Fatalf("Failed to parse url %s: %v", proxy.URL, err)
+			}
+
+			client := net.NewClient(net.Options{})
+			req, err := http.NewRequest("GET", reqURL.String(), nil)
+			if err != nil {
+				t.Error(err)
+				return
+			}
+
+			failBackend := float64(0)
+			fail := float64(0)
+			ok := float64(0)
+			var N float64
+			// iterations to make sure we have enough traffic
+			for j := 0; j < 10; j++ {
+				for i := 0; i < 1+(2*ti.windowsize); i++ {
+					until := time.Now().Add(ti.d)
```

Comment:
Maybe get rid of the loops and have
```go
// run 20x of history size
until := time.After(20 * ti.d * ti.windowsize)

work:
for {
        select {
	case <-until:
		break work
	default:
	}
...
```

Reply:
I like the nested loops because you can see that the inner makes sense and the other is "have more requests than 100"